src/goto-instrument/contracts/dynamic-frames/dfcc_instrument.cpp

```diff
@@ -361,6 +1121,87 @@ void dfcc_instrumentt::instrument_function_body(
 
   auto &body = goto_function.body;
 
+  // Builds a dependency graph representing the loop nesting structure
+  // identifies loop head and creates fresh symbols for write sets
+  grapht<loop_graph_nodet> loop_nesting_graph;
+  compute_loop_nesting_graph(function_id, loop_nesting_graph, body);
+
+  local_may_aliast local_may_alias(goto_function);
+  // Iterate over the (natural) loops in the function, in topo-sorted order,
+  // and apply any loop contracts that we find.
+  for(const auto &idx : loop_nesting_graph.topsort())
+  {
+    const auto &loop_node = loop_nesting_graph[idx];
+    if(
+      loop_node.assigns_clause.is_nil() && loop_node.invariant.is_nil() &&
+      loop_node.decreases_clause.is_nil())
+      continue;
+
+    // TODO: Fix loop contract handling for do/while loops.
+    if(
+      loop_node.latch_target->is_goto() &&
+      !loop_node.latch_target->condition().is_true())
+    {
+      log.error() << "Loop contracts are unsupported on do/while loops: "
+                  << loop_node.head_target->source_location() << messaget::eom;
+      throw 0;
+    }
+
+    const auto &loop_filter =
+      [&](const goto_programt::instructiont::targett &target)
+    { return get_loop_id_from_target(target) == idx; };
+
+    // TODO: check that no malloc and dealloc in loop body.
+    // instrument the loop body
+    instrument_instructions(
+      function_id,
+      loop_node.write_set_symbol.symbol_expr(),
+      body,
+      loop_node.head_target,
+      loop_node.latch_target,
+      cfg_info,
+      // don't skip any instructions
+      loop_filter,
+      function_pointer_contracts);
+
+    size_t outer_idx = idx;
+    if(loop_node.out.size() != 0)
+    {
+      for(const auto &i : loop_nesting_graph.topsort())
+      {
+        // Find the first successor.
+        if(loop_node.out.count(i) != 0)
+        {
+          outer_idx = i;
+          break;
+        }
+      }
+    }
+
+    const exprt &outer_write_set =
+      (outer_idx == idx)
+        ? write_set
+        : loop_nesting_graph[outer_idx].write_set_symbol.symbol_expr();
```

Comment:
I don't think this is necessarily pointing to an the enclosing loop (successor in topological order does not mean they are directly connected)